File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    let answer = n;

    for (let i = 0; i < wires.length; i++) {
        const visited = new Array(n + 1).fill(false);

        // i번째 전선 제거하고 탐색
        const count = dfs(wires, visited, wires[i][0], i);

        const diff = Math.abs(count - (n - count));
        answer = Math.min(answer, diff);
    }

    return answer;
}

function dfs(wires, visited, node, skipIndex) {
    visited[node] = true;
    let count = 1;

    for (let i = 0; i < wires.length; i++) {
        if (i === skipIndex) continue;

        const [a, b] = wires[i];

        if (a === node && !visited[b]) {
            count += dfs(wires, visited, b, skipIndex);
        } else if (b === node && !visited[a]) {
            count += dfs(wires, visited, a, skipIndex);
        }
    }

    return count;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
class Solution {
    public int solution(int n, int[][] wires) {
        int answer = n;

        for (int i = 0; i < wires.length; i++) {
            boolean[] visited = new boolean[n + 1];

            int count = dfs(wires, visited, wires[i][0], i);

            int diff = Math.abs(count - (n - count));
            answer = Math.min(answer, diff);
        }

        return answer;
    }

    private int dfs(int[][] wires, boolean[] visited, int node, int skipIndex) {
        visited[node] = true;
        int count = 1;

        for (int i = 0; i < wires.length; i++) {
            if (i == skipIndex) continue;

            int a = wires[i][0];
            int b = wires[i][1];

            if (a == node && !visited[b]) {
                count += dfs(wires, visited, b, skipIndex);
            } else if (b == node && !visited[a]) {
                count += dfs(wires, visited, a, skipIndex);
            }
        }

        return count;
    }
}